Improve Meson64 Support: `ODROID`

pyavitz opened the pull request:
Backport DTS/DTSI changes from linux-6.4.y to 6.1.y
Add meson64-reboot driver to all boards
Add board: ODROID N2L
Add uart_A uart_AO_B uart_B uart_C where appropriate
U-Boot v2023.07.02: ODROID N2/N2L/N2Plus/C4

Meson64-reboot driver: (source: tobetter)
While the current meson64-reboot driver is cleaner and doesn't depend on modding other kernel sources, its functionality leaves much to be desired. One example can be found in the ODROID C4. Using the current driver, the board will not properly power off, leaving the POWER LED still on. The new driver powers off the unit completely.

Fan control: (ODROID N2L/N2PLus)
Added service and script for controlling the trip point.
fanctrl: arguments: 65 55 45 35 25 menu run
```sh
                              ┌──┤ Fan Control ├──┐
                              │                   │
                              │    6) 65°C        │
                              │    5) 55°C        │
                              │    4) 45°C        │
                              │    3) 35°C        │
                              │    2) 25°C        │
                              │    E) Exit ..     │
                              │                   │
                              │                   │
                              │      <Ok>         │
                              │                   │
                              └───────────────────┘
```
NOTES: (N2L/HC4): I do not own the units so I can't run tests.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

igorpecovnik reviewed (approved):
Its better then current state, so lets merge.

rpardini commented:
this is a bad move for ODROID boards, and a even worse one for non-ODROID meson64's.

This is going back in many ways
1) custom fanctrl? why? this belongs in armbian-config(-ish) not here
2) hacking common kernel for shutdown?
3) changing LTS kernel with experimental stuff? (why not only in edge, at least)
4) "HACK" patches, where there used to be none?
5) removed the PWM pulse (input) reading from the HC4, why?
6) numbered patches, again, where used to be none?
7) aml-gpiomem? why?
8) changing u-boot boot order, why?

Others seem ok:
a) `n2l` addition, long overdue
b) u-boot bumps, should be ok

don't mean to offend, but this should be reverted, split into smaller PR's, and individually tested, specially on non-ODROID boards.

pyavitz commented:
I was going to give a full reply, but... Just REVERT, it is fine.

pyavitz commented:
> this is a bad move for ODROID boards, and a even worse one for non-ODROID meson64's.

How would you know, have you tried? To suggest the DTS changes are a bad move is to suggest that 6.4.y and above was a mistake on that front. Besides a few bits here and there for functionality purposes, all that was done is backport that work into LTS. The same is already being done on the BPI boards with no ill effects.

Plus this makes it so when working on the boards, all the patching is basically identical between LTS and STABLE after the patch-set is applied. It will ensure consistency. 

> 
> This is going back in many ways
> 
> 1. custom fanctrl? why? this belongs in armbian-config(-ish) not here

Is there any CURRENT armbian-config functionality on this front for the ODROIDS? I didn't see any, so I added the script and service to the two boards where it could be useful. 

> 2. hacking common kernel for shutdown?

We have already had this discussion and although I can not say with certainty these changes won't effect every board on this platform, I can say I haven't seen it happen. Without the nodes in place I don't really see how it would? But yes, further testing should be done.

> 3. changing LTS kernel with experimental stuff? (why not only in edge, at least)

Read my above thoughts on this.

> 4. "HACK" patches, where there used to be none?

Sometimes this can be useful. Don't knock it until you try it.

> 5. removed the PWM pulse (input) reading from the HC4, why?

This wasn't intentional if that is the case. What I saw looked identical to what I added to a different location of the DTS.

> 6. numbered patches, again, where used to be none?

This was done because the patching system doesn't read alphabetical the same way `ls` or the human eye would. It wasn't patching in the order I placed it, so I forced one with numbering.

> 7. aml-gpiomem? why?

Why not? Either way I'm not stuck on this. It can be removed. It was only added to LTS.

> 8. changing u-boot boot order, why?

You already do this. Why are you questioning why I would change the boot order? No offense but this one to me seems a bit odd.

> 
> Others seem ok: a) `n2l` addition, long overdue b) u-boot bumps, should be ok
> 
> don't mean to offend, but this should be reverted, split into smaller PR's, and individually tested, specially on non-ODROID boards.

No offense taken.

pyavitz commented:
@rpardini Next time before you ravage a PR, get your facts straight. I'm tired of trying to improve the situation and getting attacked for it.

https://github.com/armbian/build/blob/main/patch/u-boot/v2022.07/board_radxa-zero/0001-HACK-mmc-meson-gx-limit-to-24MHz.patch
https://github.com/armbian/build/blob/main/patch/u-boot/v2022.10/meson64-boot-usb-nvme-scsi-first.patch

Most of your argument was, at least on its face, "opinion" and not based on functionality. Try the changes next time before you take the piss.

rpardini commented:
I thought was very clear about not wanting to offend or take a piss? You of course know I highly value your many contributions, including this one. I really, really think if we had the time to review, we could have had a decent discussion and collaboration on this, instead of a merge & revert. So sorry.

> > 9\. removed the PWM pulse (input) reading from the HC4, why?
> 
> This wasn't intentional if that is the case. What I saw looked identical to what I added to a different location of the DTS.

- this was due to the `boards-` multi-patches, which group a bunch of unrelated changes to multiple boards _and_ common dtsi's. let's... not do that?
  - We had a hard-earned, multi-year, multi-person effort, decent-ish organization with (mostly) each board (`board-`) being patched in it's own file. 
  - Due to this, I missed the fact you moved the HC4's `pulses-per-revolutions` et al there. So yeah, it might very well have worked.

> Plus this makes it so when working on the boards, all the patching is basically identical between LTS and STABLE after the patch-set is applied. It will ensure consistency.

- I don't think we'll find common ground here. LTS is LTS for a reason; if we go backporting-all-the-things, LTS loses all meaning. \<sarcasm>For better consistency, why not bump both `current` and `edge` to `for-next`?\</sarcasm> 
- What happens when an eventual fix on say 6.1.123 doesn't apply anymore due to the backports?

> > 13\. aml-gpiomem? why?
> 
> Why not? Either way I'm not stuck on this. It can be removed. It was only added to LTS.

Sorry -- that makes no sense at all -- it should be added to `edge`, tested, and maybe included (probably by accident) in the _next_ LTS if proven stable, useful, and easy to maintain... adding it "only to LTS" would only make sure it is dropped eventually.

Regarding "why not", is there at least a use case for this? (I think I saw CoreELEC carrying similar on S4/T7 for some reason). Case in point: could we switch meson64 to a `-rt` kernel and say "why not"...

> > 7\. "HACK" patches, where there used to be none?
> 
> Sometimes this can be useful. Don't knock it until you try it.

Well exactly like you pointed out, I've myself added HACK patches in the past. But only after a lot of public discussion and looking for alternatives, and somewhat-agreeing with upstream. Here you added it to `odroidn2`, which never had it, without any justification... is there any? (_I can only speculate it's because you bumped it to UHS (which was a much heated discussion point since 2020), added the reboot driver, hacked the shutdown, and the HACK patch is the magic-tie-off?_)

> > 15\. changing u-boot boot order, why?
> 
> You already do this. Why are you questioning why I would change the boot order? No offense but this one to me seems a bit odd.

Yes, I started this trend, for a simple reason: having USB, NVMe and SCSI boot first, especially for a bootloader deployed to MTD, allows for (non-Armbian, even) generic (possibly EFI) booting. So I'm not questioning the "change of order", but instead the "actual order" and the fact the patch is not shared across boards.

- My question was why you decided to use a different-different order? (I already had "USB/NVMe/SCSI", you changed to "USB", and on boards with no NVMe/SCSI, having them higher up would make no difference).
- Also at the 22.10, 23.01, and 23.07 level, the patch sat at the root: so _all_ meson64 boards adopting that u-boot version would get the same ordering, and a single patch, instead of duplicating it across `board_` directories. 
  - That was meant as a feature: "all Armbian meson64 boards using 22.10+ u-boot boot from USB/NVMe/SCSI first."

That said, yeah, sorry, I should definitely have explained better.

> Most of your argument was, at least on its face, "opinion" and not based on functionality. Try the changes next time before you take the piss.

I believe you wrote this in frustration and anger. I'm sorry. I hope to have at least convinced you I didn't do that. Again, I'm sure if we had more time to discuss pre-merge, we would have found a way to include almost everything in here, just done in slightly different ways.

rpardini commented:
One last one. N2L doesn't have detection code (via ADC) in uboot, so it would either fail to boot or boot with a wrong DTB.

rpardini commented:
Sorry, one more. `meson-g12b.conf` matched `[[ $BOARD == odroidn2* ]]` for the FIP blobs.
The N2L, though, _LPDDR4 RAM is 20% faster than the N2+. The N2L’s LPDDR4 RAM is running at 1608Mhz while N2’s DDR4 was running at 1320Mhz_ -- thus requiring different blobs. 
Blobs for the N2L were added to https://github.com/hardkernel/u-boot/commit/154ddfc826 and sit in the `odroid-n2l` folder in amlogic-boot-fip.

pyavitz commented:
> I thought was very clear about not wanting to offend or take a piss? You of course know I highly value your many contributions, including this one. I really, really think if we had the time to review, we could have had a decent discussion and collaboration on this, instead of a merge & revert. So sorry.
> 
> > > 9. removed the PWM pulse (input) reading from the HC4, why?
> > 
> > 
> > This wasn't intentional if that is the case. What I saw looked identical to what I added to a different location of the DTS.
> 
> * this was due to the `boards-` multi-patches, which group a bunch of unrelated changes to multiple boards _and_ common dtsi's. let's... not do that?
>   
>   * We had a hard-earned, multi-year, multi-person effort, decent-ish organization with (mostly) each board (`board-`) being patched in it's own file.
>   * Due to this, I missed the fact you moved the HC4's `pulses-per-revolutions` et al there. So yeah, it might very well have worked.
> 
> > Plus this makes it so when working on the boards, all the patching is basically identical between LTS and STABLE after the patch-set is applied. It will ensure consistency.
> 
> * I don't think we'll find common ground here. LTS is LTS for a reason; if we go backporting-all-the-things, LTS loses all meaning. <sarcasm>For better consistency, why not bump both `current` and `edge` to `for-next`?</sarcasm>
> * What happens when an eventual fix on say 6.1.123 doesn't apply anymore due to the backports?

Fair enough, but that can happen now without my changes. Let us not make things better and improve upon the original ground work because "a" "b" or "c" can happen? By this logic we shouldn't change anything.

Whilst working with the BPI patches I noticed for whatever reason there were patches being applied to 6.1 and 6.4 that did the same thing but were being applied to different locations of the DTS or DTSI. This a problem if you are trying to add features or make adjustments to the DTS files across versions.

> 
> > > 13. aml-gpiomem? why?
> > 
> > 
> > Why not? Either way I'm not stuck on this. It can be removed. It was only added to LTS.
> 
> Sorry -- that makes no sense at all -- it should be added to `edge`, tested, and maybe included (probably by accident) in the _next_ LTS if proven stable, useful, and easy to maintain... adding it "only to LTS" would only make sure it is dropped eventually.
> 
> Regarding "why not", is there at least a use case for this? (I think I saw CoreELEC carrying similar on S4/T7 for some reason). Case in point: could we switch meson64 to a `-rt` kernel and say "why not"...

Well it is a feature. It isn't honestly one I ever use, but could be useful. Adding it to LTS which is the kernel most people should be using would provide those the opportunity to use it. Anyway, like I said, I'm not stuck on it and it can be removed. One less headache.

> > > 7. "HACK" patches, where there used to be none?
> > 
> > 
> > Sometimes this can be useful. Don't knock it until you try it.
> 
> Well exactly like you pointed out, I've myself added HACK patches in the past. But only after a lot of public discussion and looking for alternatives, and somewhat-agreeing with upstream. Here you added it to `odroidn2`, which never had it, without any justification... is there any? (_I can only speculate it's because you bumped it to UHS (which was a much heated discussion point since 2020), added the reboot driver, hacked the shutdown, and the HACK patch is the magic-tie-off?_)

This debate over the reboot and power down issue on these boards has been beaten to death already. The patches I included in the PR "HACK, meson64-reboot and DTS changes" in my experience are a better take on the solution than the one currently in use by Armbian.

If there is a reason for not wanting to use it, that is fine.

> 
> > > 15. changing u-boot boot order, why?
> > 
> > 
> > You already do this. Why are you questioning why I would change the boot order? No offense but this one to me seems a bit odd.
> 
> Yes, I started this trend, for a simple reason: having USB, NVMe and SCSI boot first, especially for a bootloader deployed to MTD, allows for (non-Armbian, even) generic (possibly EFI) booting. So I'm not questioning the "change of order", but instead the "actual order" and the fact the patch is not shared across boards.
> 
> * My question was why you decided to use a different-different order? (I already had "USB/NVMe/SCSI", you changed to "USB", and on boards with no NVMe/SCSI, having them higher up would make no difference).
> * Also at the 22.10, 23.01, and 23.07 level, the patch sat at the root: so _all_ meson64 boards adopting that u-boot version would get the same ordering, and a single patch, instead of duplicating it across `board_` directories.
>   
>   * That was meant as a feature: "all Armbian meson64 boards using 22.10+ u-boot boot from USB/NVMe/SCSI first."
> 
> That said, yeah, sorry, I should definitely have explained better.

That is because, blindly applying that boot order to every board on the platform is a bad idea. So I isolated them and only included the relevant part of the patch. This NVMe and SCSI bit doesn't need to be there on these boards.

> 
> > Most of your argument was, at least on its face, "opinion" and not based on functionality. Try the changes next time before you take the piss.
> 
> I believe you wrote this in frustration and anger. I'm sorry. I hope to have at least convinced you I didn't do that. Again, I'm sure if we had more time to discuss pre-merge, we would have found a way to include almost everything in here, just done in slightly different ways.

Annoyed not angry. The changes I made were not just blind changes. I use them on my units. If they didn't work or make things better, I wouldn't have created the PR in the first place. All I'm really saying is maybe try the changes next time before accusing it of being "the worst thing to happen to cats, since dogs".

pyavitz commented:
> Well exactly like you pointed out, I've myself added HACK patches in the past. But only after a lot of public discussion and looking for alternatives, and somewhat-agreeing with upstream. Here you added it to odroidn2, which never had it, without any justification... is there any? (I can only speculate it's because you bumped it to UHS (which was a much heated discussion point since 2020), added the reboot driver, hacked the shutdown, and the HACK patch is the magic-tie-off?)

As for the N2, there is no harm here. Adding or not adding the HACK doesn't really seem to matter much, but it also doesn't hurt. It is more of a "let us cover all our bases" type thing. The worst thing you can do to the N2 to create instability issues is overlock the board beyond `Cortex-A53 @ 1992MHz 40°C, Cortex-A73 @ 1908MHz 41°C`. If you go any higher you will get random lockups and if watchdog is in place, reboots.

hzyitc commented:
I think we should have the ablity to build an PR and do tests with our auto test framework when our member required. This will be benefit when try to merging for such large family.

pyavitz commented:
> Regarding "why not", is there at least a use case for this? (I think I saw CoreELEC carrying similar on S4/T7 for some reason). Case in point: could we switch meson64 to a -rt kernel and say "why not"...

Sorry, I was not going to dig into this statement and some of the others I have seen, but this one is especially stupid. You don't know anything about `-rt` kernels and my knowledge of them is some what limited as well, but I do know they are very version and commit specific. You aren't going to just support them wiliy-nilly in Armbian and keep upping the kernel release using the same patch set. The aml-gpiomen patch is nothing like `-rt` and is also `node` specific. This would only have shown up in an Odroid.

You keep making pointless statements to justify your poor attitude towards other contributors.

igorpecovnik commented:
> I think we should have the ablity to build an PR and do tests with our auto test framework when our member required. 

We are not that far away from having that. https://armbian.atlassian.net/browse/AR-1825

rpardini commented:
> The patches I included in the PR "HACK, meson64-reboot and DTS changes" in my experience are a better take on the solution than the one currently in use by Armbian. If there is a reason for not wanting to use it, that is fine.

The reason is simple: the changes are not contained. We spent months de-ODROID'ying g12a/b so we could bring in other, non-ODROID boards into the family. That involved removal of the shutdown revert patch (that last I looked, is still in tobetter's tree, and yes, "it works" on ODROID.) 

Years later, someone re-added the meson64 reboot patch, by saying they "fixed the reboot on C4". It didn't, clearly? You did changes to such stuff as `arch/arm64/include/asm/system_misc.h` and `arch/arm64/kernel/process.c` -- would you ever send those upstream, or have any justification at all?

Maybe you'll remember the clock phase fiasco? Remember how much we talked, how  many people were involved, and even upstream, before settling on an approach? That's collaboration -- sending _moar hacks_ is not. Sorry.

> Well it is a feature. It isn't honestly one I ever use, but could be useful.

Ok so why did you add it, then? 🤷 

> blindly applying that boot order to every board on the platform is a bad idea

yes. except _nothing_ was done blindly, and definitely not to every board. 

> but this one is especially stupid

thanks! but...

> Yes, it would just fail to boot.

oops... so I'm not that stupid then?

> justify your poor attitude towards other contributors.

well, oops again. I apologized from the start, justified _every single one_ of my objections, etc.
you call me stupid after getting a review and it is _me_ who has an attitude problem?

> close this thread

deal.
